client/verta/verta/runtime.py

```diff
@@ -0,0 +1,220 @@
+# -*- coding: utf-8 -*-
+"""Classes and functions to enable context logging within a model's predict() function.
+.. versionadded:: 0.22.0
+
+"""
+
+import json
+import re
+import threading
+from typing import Any, Dict, Optional
+
+
+_THREAD = threading.local()
+_S3_REGEX = re.compile("[0-9a-zA-Z_-]+$")
+
+
+def _init_thread_logs() -> None:
+    """
+    Initialize our thread-local variable for storing logging context.
+    """
+    if not hasattr(_THREAD, 'logs'):
+        _THREAD.logs = dict()
+
+
+def _get_thread_logs() -> Dict[str, Any]:
+    """
+    Return the current thread-local logs or initialize a new dict.
+    """
+    _init_thread_logs()
+    return _THREAD.logs
+
+
+def _set_thread_logs(logs: Dict[str, Any]) -> Dict[str, Any]:
+    """
+    Sets the thread-local logs, overwriting any existing values.
+    """
+    _THREAD.logs = logs
+    return _THREAD.logs
+
+
+def _init_validate_flag() -> None:
+    """
+    Initialize our thread-local variable for the validation flag.
+    """
+    if not hasattr(_THREAD, 'validate'):
+        _THREAD.validate = False
+
+
+def _get_validate_flag() -> bool:
+    """
+    Return the current thread-local variable for validate or initialize a
+    new boolean.
+    """
+    _init_validate_flag()
+    return _THREAD.validate
+
+
+def _set_validate_flag(flag: bool) -> bool:
+    """
+    Sets the thread-local validate flag boolean value.
+    """
+    _THREAD.validate = flag
+    return _THREAD.validate
+
+
+def _validate_json(value: Any) -> str:
+    """
+    Attempt to serialize the provided value to JSON.
+    Raise an error on failure.
+    """
+    try:
+        return json.dumps(value)
+    except TypeError as type_err:
+        print("provided data type cannot be converted to JSON for logging.")
+        raise type_err
+
+
+def _validate_s3(
+        value: str,
+        pattern: re.Pattern = _S3_REGEX
+    ) -> None:
+    """
+    Check the provided value for any special characters that would
+    violate Amazon S3 object key naming rules. Raise ValueError if any
+    are found.
+    """
+    if len(value) > 100:
```

Comment:
I used `timeit` to measure the execution time of Python's native `len()` function versus adding the length check to the regex like `[0-9a-zA-Z_-]{1,100}$`.  The difference was very negligible and both are pretty fast, so I went with the Python length check so I could throw a more informative error message back to the user in that case.